cpp/meta_module/algorithm/meta.cpp

```diff
@@ -0,0 +1,284 @@
+
+  int64_t get_label_count() const { return static_cast<int64_t>(labels.size()); }
+  int64_t get_relationship_type_count() const { return static_cast<int64_t>(relationship_types_cnt.size()); }
+  int64_t get_property_key_count() const { return static_cast<int64_t>(property_key_cnt.size()); }
+
+  void reset();
+  void update_labels(const mgp::Node &node, int add);
+  void update_property_key_cnt(const mgp::Node &node, int add);
+  void update_property_key_cnt(const mgp::Relationship &relationship, int add);
+  void update_relationship_types(const mgp::Relationship &relationship, int add);
+  void update_relationship_types_cnt(const mgp::Relationship &relationship, int add);
+};
+
+// Global variable
+// This is unsafe in the multithreaded environment, the workaround would be building a thread-safe dynamic storage
+// implementation
+Metadata metadata;
+
+void insert(std::unordered_map<std::string, int64_t> &map, std::string &key, int64_t add) {
+  auto iterator = map.find(key);
+  if (iterator != map.end()) {
+    (*iterator).second += add;
+    if ((*iterator).second == 0) {
+      map.erase(iterator);
+    }
+  } else {
+    map[key] = add;
+  }
+}
+
+void insert(std::unordered_map<std::string, int64_t> &map, std::string_view key_view, int64_t add) {
+  auto key = std::string(key_view);
+  insert(map, key, add);
+}
+
+void Metadata::reset() {
+  node_cnt = relationship_cnt = 0;
+  labels.clear();
+  property_key_cnt.clear();
+  relationship_types.clear();
+  relationship_types_cnt.clear();
+}
+
+void Metadata::update_labels(const mgp::Node &node, int add) {
+  for (const auto &label : node.Labels()) {
+    insert(labels, label, add);
+  }
+}
+
+void Metadata::update_property_key_cnt(const mgp::Node &node, int add) {
+  for (const auto &[property, _] : node.Properties()) {
+    insert(property_key_cnt, property, add);
+  }
+}
+
+void Metadata::update_property_key_cnt(const mgp::Relationship &relationship, int add) {
+  for (const auto &[property, _] : relationship.Properties()) {
+    insert(property_key_cnt, property, add);
+  }
+}
+
+void Metadata::update_relationship_types(const mgp::Relationship &relationship, int add) {
+  const auto type = relationship.Type();
+
+  for (const auto &label : relationship.From().Labels()) {
+    auto key = "(:" + std::string(label) + ")-[:" + std::string(type) + "]->()";
+    insert(relationship_types, key, add);
+  }
+  for (const auto &label : relationship.To().Labels()) {
+    auto key = "()-[:" + std::string(type) + "]->(:" + std::string(label) + ")";
+    insert(relationship_types, key, add);
+  }
+
+  insert(relationship_types, "()-[:" + std::string(type) + "]->()", add);
+}
+
+void Metadata::update_relationship_types_cnt(const mgp::Relationship &relationship, int add) {
+  insert(relationship_types_cnt, relationship.Type(), add);
+}
+
+Metadata get_all_metadata(mgp_graph *memgraph_graph) {
+  Metadata data;
+
+  const mgp::Graph graph{memgraph_graph};
+
+  for (const auto node : graph.Nodes()) {
```

Comment:
&node

Reply:
Actually doesn't need &  here this is actually copied under the hood.